pkgs/misc/drivers/xow/default.nix

```
@@ -0,0 +1,31 @@
+{ stdenv, fetchFromGitHub, libusb, gitMinimal }:
+
+stdenv.mkDerivation rec {
+  pname = "xow";
+  version = "0.2";
+
+  src = fetchFromGitHub {
+    owner = "medusalix";
+    repo = "xow";
+    rev = "v${version}";
+    sha256 = "03ajal91xi52svzy621aa4jcdf0vj4pqd52kljam0wryrlmcpbr3";
+  };
+
+  makeFlags = "BUILD=RELEASE";
```

```suggestion
  makeFlags = [
    "INSTALL_PATH=${placeholder ''out''}/bin"
    "SERVICE_PATH=${placeholder ''out''}/lib/systemd/user"
    "BUILD=RELEASE"
  ];
```

will allow you to use the default `installPhase`.

---

Sadly, not really:

```
cp: cannot create regular file '/nix/store/jd1mrvsb21pcxn4s1x6681rkypybj455-xow-0.2/lib/systemd/user': No such file or directory
```

I'm not sure about fixing this, though. Should the xow author do this in the `Makefile`?

---

Apparently, it is not trying to create the dirctetory. You will also need to add `mkdir -p $out/lib/systemd/user` into `preInstall`.

---

Hmm, apparently, the install target also tries to enable and start the service, which is an incredibly bad practice:

https://github.com/medusalix/xow/blob/897eee07351f88a04eb16a02b8bca0b91f91cad4/Makefile#L32-L33

---

Hi, developer here. Could you please explain why this is a bad practice?

---

Hello. The main issue is that the `install` target is, by convention, supposed to install things, not reconfigure the init system or start the installed software. As such it is at odds with the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Another issue is that most modern Linux distributions actually build and install packages in sandboxed environment (chroot), often locked out so it might not even be able to call `systemctl`. And then they take a snapshot of the created directory tree and ship it as a tarball to users. The user’s package manager will install the package by just extracting the archive over the file system (possibly running some distribution specific hooks), so the service would not be started for the user anyway.

---

Thanks for the explanation. I've fixed some of the issues in the [`install`](https://github.com/medusalix/xow/blob/install/Makefile) branch, though I'm still unsure about the udev rules. We've basically got three options:

1. Make uinput and USB devices r/w for all users (current option).
2. Make uinput and USB devices r/w for current user (udev tag `uaccess`). What do we do in the `service` file? Only specify a single user?
3. Make uinput and USB devices r/w for a single group (could be named `xow`).

What do you think? What's the best option (both for NixOS and for users building from source)?

---

I am not a udev expert but I have seen `uaccess` recommended over `GROUP` and `MODE` in other PRs. Though apparently, the filename must be ordered before `73-seat-late.rules` according to https://github.com/systemd/systemd/issues/4288#issuecomment-348166161. As for the service, would it be possible to make it a user service? That way, it would already run under `plugdev` group (I understand is required for udev access).

---

As far as I understand, with user services we would get a service instance of `xow` for each user session that is currently active. That's quite problematic because only one instance can have the USB device open at the same time. I'm now using systemd's `DynamicUser` (which should be quite safe), but that doesn't work with `uaccess`.

---

Would [`SupplementaryGroups=plugdev`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#SupplementaryGroups=) work? Otherwise, I would try writing to [systemd ML](https://lists.freedesktop.org/mailman/listinfo/systemd-devel).

---

It certainly works, but it's limited to Ubuntu and Debian and is deprecated according to [`systemd`](https://github.com/systemd/systemd/issues/5802#issuecomment-297026399).